src/printer.rs

```diff
@@ -160,78 +264,94 @@ impl Printer {
+                    self.buffer.print("\n")?;
                 }
             }
-        };
+        }
         Ok(())
     }
 
-    pub async fn print_response_body(&mut self, mut response: Response) -> Result<()> {
-        match get_content_type(&response.headers()) {
-            Some(ContentType::Json) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_json(&String::from_utf8_lossy(&bytes))?;
+    pub fn print_response_body(&mut self, mut response: Response) -> anyhow::Result<()> {
+        if !self.buffer.is_terminal() {
+            // No trailing newlines, no decoding, direct streaming
+            self.print_body_stream(get_content_type(&response.headers()), &mut response)?;
+        } else if self.stream {
+            match self.print_body_stream(
+                get_content_type(&response.headers()),
+                &mut decode_stream(&mut response),
+            ) {
+                Ok(_) => {
+                    self.buffer.print("\n")?;
                 }
-            }
-            Some(ContentType::Xml) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_xml(&String::from_utf8_lossy(&bytes))?;
-                }
-            }
-            Some(ContentType::Html) if self.stream => {
-                while let Some(bytes) = response.chunk().await? {
-                    self.print_html(&String::from_utf8_lossy(&bytes))?;
-                }
-            }
-            Some(ContentType::Json) => self.print_json(&response.text().await?)?,
-            Some(ContentType::Xml) => self.print_xml(&response.text().await?)?,
-            Some(ContentType::Html) => self.print_html(&response.text().await?)?,
-            _ => {
-                let mut is_first_chunk = true;
-                while let Some(bytes) = response.chunk().await? {
-                    if is_first_chunk
-                        && matches!(self.buffer, Buffer::Stdout | Buffer::Stderr)
-                        && bytes.contains(&b'\0')
-                    {
-                        self.buffer.print(BINARY_SUPPRESSOR)?;
-                        break;
+                Err(err) if err.kind() == io::ErrorKind::InvalidData => {
+                    if self.color {
+                        self.buffer.print("\x1b[0m")?;
                     }
-                    is_first_chunk = false;
-
-                    self.buffer.write_all(&bytes)?;
+                    self.buffer.print(BINARY_SUPPRESSOR)?;
                 }
+                Err(err) => return Err(err.into()),
+            }
+        } else {
+            let content_type = get_content_type(&response.headers());
+            // Note that .text() behaves like String::from_utf8_lossy()
+            let text = response.text()?;
+            if text.contains('\0') {
+                self.buffer.print(BINARY_SUPPRESSOR)?;
+                return Ok(());
```

Comment:
What would happen if we used `BinaryGuard` for non-streamed responses?

Reply:
Okay, so `BinaryGuard` is only necessary when we are streaming a response but for non-streamed responses, we can just scan the whole text?

Reply:
Yep, `BinaryGuard` does the same detection incrementally.

Reply:
Can you add that as a comment to `BinaryGuard`?